tests/integrations/raftstore/test_unsafe_recovery.rs

```diff
@@ -0,0 +1,535 @@
+    assert_eq!(cluster.must_get(b"k4"), Some(b"v4".to_vec()));
+}
+
+#[test]
+fn test_force_leader_for_learner() {
+    let mut cluster = new_node_cluster(0, 5);
+    cluster.pd_client.disable_default_operator();
+
+    cluster.run();
+    cluster.must_put(b"k1", b"v1");
+
+    let region = cluster.get_region(b"k1");
+    cluster.must_split(&region, b"k9");
+    let mut region = cluster.get_region(b"k2");
+    let peer_on_store5 = find_peer(&region, 5).unwrap();
+    cluster.must_transfer_leader(region.get_id(), peer_on_store5.clone());
+
+    let peer_on_store1 = find_peer(&region, 1).unwrap();
+    // replace one peer with learner
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), peer_on_store1.clone());
+    cluster.pd_client.must_add_peer(
+        region.get_id(),
+        new_learner_peer(peer_on_store1.get_store_id(), peer_on_store1.get_id()),
+    );
+
+    must_get_equal(&cluster.get_engine(1), b"k1", b"v1");
+
+    cluster.stop_node(3);
+    cluster.stop_node(4);
+    cluster.stop_node(5);
+
+    let put = new_put_cmd(b"k2", b"v2");
+    let req = new_request(region.get_id(), region.take_region_epoch(), vec![put], true);
+    // majority is lost, can't propose command successfully.
+    assert!(
+        cluster
+            .call_command_on_leader(req, Duration::from_millis(10))
+            .is_err()
+    );
+
+    cluster.enter_force_leader(region.get_id(), 1, HashSet::from_iter(vec![3, 4, 5]));
+    // promote the learner first and remove the peers on failed nodes
+    cluster
+        .pd_client
+        .must_add_peer(region.get_id(), find_peer(&region, 1).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 3).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 4).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 5).unwrap().clone());
+    cluster.exit_force_leader(region.get_id(), 1);
+
+    // majority is formed, can propose command successfully now
+    cluster.must_put(b"k4", b"v4");
+    assert_eq!(cluster.must_get(b"k2"), None);
+    assert_eq!(cluster.must_get(b"k3"), None);
+    assert_eq!(cluster.must_get(b"k4"), Some(b"v4".to_vec()));
+    cluster.must_transfer_leader(region.get_id(), find_peer(&region, 1).unwrap().clone());
+}
+
+#[test]
+fn test_force_leader_on_healthy_region() {
+    let mut cluster = new_node_cluster(0, 5);
+    cluster.pd_client.disable_default_operator();
+
+    cluster.run();
+    cluster.must_put(b"k1", b"v1");
+
+    let region = cluster.get_region(b"k1");
+    cluster.must_split(&region, b"k9");
+    let region = cluster.get_region(b"k2");
+    let peer_on_store5 = find_peer(&region, 5).unwrap();
+    cluster.must_transfer_leader(region.get_id(), peer_on_store5.clone());
+
+    // try to enter force leader, it can't succeed due to majority isn't lost
```

Comment:
Where is it checked?

Reply:
`check_force_leader` will fail and exit pre force leader

Reply:
But the peer will start campaign any way. It can break correctness as we assume it should not campaign if there is still leader.

Reply:
note that `promotable` is still false after being force leader, so it will not start campaign.

Reply:
How can it be related to `promotable`?

Reply:
sorry, misread with the below comment. forget about it.

`It can break correctness as we assume it should not campaign if there is still leader.`
Yes, we don't use `prevote` now, so it can't avoid interrupting other peers if there is a leader.